File: substrate/research_wrestle_competition_quality_compose.py
"""Research wrestle + competition quality supercompose (pure).

live_dispatch_authorized, remote_fetched, backlog_mutated always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.competition_dr_quality_source_pack_compose import (
    CompetitionDrQualitySourcePackCompose,
    CompetitionDrQualitySourcePackComposeError,
    compose_competition_dr_quality_source_pack,
)
from substrate.research_wrestle_session_supercompose import (
    ResearchWrestleSessionSupercompose,
    ResearchWrestleSessionSupercomposeError,
    compose_research_wrestle_session,
)


class ResearchWrestleCompetitionQualityComposeError(ValueError):
    """Fail-closed validation for wrestle + competition quality pack."""


@dataclass(frozen=True)
class ResearchWrestleCompetitionQualityCompose:
    session_id: str
    parent_asset_id: str
    wrestle: ResearchWrestleSessionSupercompose
    competition_quality: CompetitionDrQualitySourcePackCompose
    session_ready: bool
    live_dispatch_authorized: bool
    remote_fetched: bool
    backlog_mutated: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "parent_asset_id": self.parent_asset_id,
            "wrestle": self.wrestle.to_dict(),
            "competition_quality": self.competition_quality.to_dict(),
            "session_ready": self.session_ready,
            "live_dispatch_authorized": False,
            "remote_fetched": False,
            "backlog_mutated": False,
            "notes": list(self.notes),
            "authority": "research_wrestle_competition_quality_compose_advisory",
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise ResearchWrestleCompetitionQualityComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_research_wrestle_competition_quality(
    *,
    session_id: object,
    parent_asset_id: object,
    floating_instance_count: object,
    completed_floating_count: object,
    twin_insight_count: object,
    twin_question_count: object,
    open_question_count: object,
    competitor_decisions: object,
    requested_families: object,
    citations: object,
    quality_overall: object,
    would_exceed: object,
    operator_ack: object,
    focus_areas: object | None = None,
    filter_to_selected_families: object | None = None,
    quality_floor: object | None = None,
    preferred_view_mode: object | None = None,
    operator_override: object | None = None,
    require_no_behind_gaps: object | None = None,
) -> ResearchWrestleCompetitionQualityCompose:
    """Compose wrestle session + competition quality pack. Never dispatches."""
    if not isinstance(operator_ack, bool):
        raise ResearchWrestleCompetitionQualityComposeError(
            "operator_ack must be an explicit boolean"
        )
    session = _require_nonempty(session_id, field="session_id")
    parent = _require_nonempty(parent_asset_id, field="parent_asset_id")

    notes: list[str] = [
        "live_dispatch_authorized=false — wrestle+quality pack is pure readiness",
        "remote_fetched=false — no arxiv/substack network fetch",
        "backlog_mutated=false — competition residuals advisory only",
    ]

    try:
        competition_quality = compose_competition_dr_quality_source_pack(
            session_id=session,
            competitor_decisions=competitor_decisions,
            requested_families=requested_families,
            citations=citations,
            quality_overall=quality_overall,
            would_exceed=would_exceed,
            operator_ack=operator_ack,
            focus_areas=focus_areas,
            filter_to_selected_families=filter_to_selected_families,
            quality_floor=quality_floor,
            operator_override=operator_override,
            require_no_behind_gaps=require_no_behind_gaps,
        )
    except CompetitionDrQualitySourcePackComposeError as e:
        raise ResearchWrestleCompetitionQualityComposeError(str(e)) from e
    notes.extend(competition_quality.notes)

    families = competition_quality.citations.selection.families
    family_count = len(families) if families else (
        len(requested_families) if isinstance(requested_families, list) else 0
    )

    try:
        wrestle = compose_research_wrestle_session(
            session_id=session,
            parent_asset_id=parent,
            floating_instance_count=floating_instance_count,
            completed_floating_count=completed_floating_count,
            twin_insight_count=twin_insight_count,
            twin_question_count=twin_question_count,
            open_question_count=open_question_count,
            source_family_count=family_count,
            citation_pack_ready=competition_quality.citations.pack_ready,
            quality_overall=quality_overall,
            would_exceed=would_exceed,
            quality_floor=quality_floor,
            preferred_view_mode=preferred_view_mode,
            operator_override=operator_override,
        )
    except ResearchWrestleSessionSupercomposeError as e:
        raise ResearchWrestleCompetitionQualityComposeError(str(e)) from e
    notes.extend(wrestle.notes)

    session_ready = (
        wrestle.wrestle_ready
        and competition_quality.pack_ready
        and wrestle.live_dispatch_authorized is False
        and competition_quality.live_dispatch_authorized is False
    )
    if not wrestle.wrestle_ready:
        notes.append("session_ready=false — wrestle substrate not ready")
    elif not competition_quality.pack_ready:
        notes.append(
            "session_ready=false — competition/quality/source pack not ready"
        )
    else:
        notes.append(
            "session_ready=true — wrestle+competition quality intent only; still pure"
        )

    if (
        wrestle.live_dispatch_authorized is not False
        or competition_quality.live_dispatch_authorized is not False
        or competition_quality.remote_fetched is not False
        or competition_quality.backlog_mutated is not False
    ):
        raise ResearchWrestleCompetitionQualityComposeError(
            "invariant: nested honesty flags must remain false"
        )

    notes.extend(
        (
            "live_dispatch_authorized=false",
            "remote_fetched=false",
            "backlog_mutated=false",
        )
    )

    return ResearchWrestleCompetitionQualityCompose(
        session_id=session,
        parent_asset_id=parent,
        wrestle=wrestle,
        competition_quality=competition_quality,
        session_ready=session_ready,
        live_dispatch_authorized=False,
        remote_fetched=False,
        backlog_mutated=False,
        notes=tuple(notes),
        authority="research_wrestle_competition_quality_compose_advisory",
    )


def format_research_wrestle_competition_quality_summary(
    c: ResearchWrestleCompetitionQualityCompose,
) -> str:
    return (
        f"session_ready={c.session_ready} · "
        f"wrestle_ready={c.wrestle.wrestle_ready} · "
        f"quality_pack_ready={c.competition_quality.pack_ready} · "
        f"citations={c.competition_quality.citations.citation_count} · "
        f"behind={c.competition_quality.competition.behind_count} · "
        f"live_dispatch_authorized=false · remote_fetched=false · "
        f"backlog_mutated=false"
    )


__all__ = [
    "ResearchWrestleCompetitionQualityCompose",
    "ResearchWrestleCompetitionQualityComposeError",
    "compose_research_wrestle_competition_quality",
    "format_research_wrestle_competition_quality_summary",
]

File: apps/reading/src/api/researchWrestleCompetitionQualityCompose.ts
/**
 * Research wrestle + competition quality supercompose (pure).
 *
 * Operator vision: live in the research workstation wrestling with information
 * while holding the highest-quality deep research bar — competition gaps,
 * arxiv/substack citations, quality/budget honesty — without live dispatch.
 *
 * live_dispatch_authorized always false.
 * remote_fetched always false.
 * backlog_mutated always false.
 */

import {
  composeResearchWrestleSession,
  type ResearchWrestleSessionInput,
  type ResearchWrestleSessionSupercompose,
} from "./researchWrestleSessionSupercompose";
import {
  composeCompetitionDrQualitySourcePack,
  type CompetitionDrQualitySourcePackCompose,
  type CompetitionDrQualitySourcePackInput,
} from "./competitionDrQualitySourcePackCompose";
import type { CompetitorDecision, DecisionArea } from "./competitionDeepResearchGap";
import type { CitationFamily, CitationRecord } from "./deepResearchSourceCitationPack";

export interface ResearchWrestleCompetitionQualityInput {
  session_id: string;
  parent_asset_id: string;
  /** Wrestle substrate signals (caller-supplied counts/flags). */
  floating_instance_count: number;
  completed_floating_count: number;
  twin_insight_count: number;
  twin_question_count: number;
  open_question_count: number;
  preferred_view_mode?: "floating" | "fullscreen" | null;
  /** Competition + citation + quality/budget inputs. */
  competitor_decisions: CompetitorDecision[];
  focus_areas?: DecisionArea[] | null;
  requested_families: CitationFamily[];
  citations: CitationRecord[];
  filter_to_selected_families?: boolean;
  quality_overall: number | null;
  quality_floor?: number;
  would_exceed: boolean | null;
  operator_override?: boolean;
  require_no_behind_gaps?: boolean;
  operator_ack: boolean;
}

export interface ResearchWrestleCompetitionQualityCompose {
  session_id: string;
  parent_asset_id: string;
  wrestle: ResearchWrestleSessionSupercompose;
  competition_quality: CompetitionDrQualitySourcePackCompose;
  /**
   * True when wrestle_ready and competition_quality.pack_ready.
   * Still never live-dispatches.
   */
  session_ready: boolean;
  live_dispatch_authorized: false;
  remote_fetched: false;
  backlog_mutated: false;
  notes: string[];
  authority: "research_wrestle_competition_quality_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Super-compose wrestle session + world-class DR competition/quality pack.
 * Never live-dispatches; never scrapes; never mutates backlog.
 */
export function composeResearchWrestleCompetitionQuality(
  input: ResearchWrestleCompetitionQualityInput,
): ResearchWrestleCompetitionQualityCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  const session_id = requireNonEmpty(input.session_id, "session_id");
  const parent_asset_id = requireNonEmpty(
    input.parent_asset_id,
    "parent_asset_id",
  );

  const notes: string[] = [
    "live_dispatch_authorized=false — wrestle+quality pack is pure readiness",
    "remote_fetched=false — no arxiv/substack network fetch",
    "backlog_mutated=false — competition residuals advisory only",
  ];

  // Derive source_family_count and citation_pack_ready from competition pack
  // after compose; first pass quality pack, then wrestle with its signals.
  const qualityInput: CompetitionDrQualitySourcePackInput = {
    session_id,
    competitor_decisions: input.competitor_decisions,
    focus_areas: input.focus_areas,
    requested_families: input.requested_families,
    citations: input.citations,
    filter_to_selected_families: input.filter_to_selected_families,
    quality_overall: input.quality_overall,
    quality_floor: input.quality_floor,
    would_exceed: input.would_exceed,
    operator_override: input.operator_override,
    operator_ack: input.operator_ack,
    require_no_behind_gaps: input.require_no_behind_gaps,
  };
  const competition_quality =
    composeCompetitionDrQualitySourcePack(qualityInput);
  notes.push(...competition_quality.notes);

  const familyCount = Array.isArray(
    competition_quality.citations.selection?.families,
  )
    ? competition_quality.citations.selection.families.length
    : Array.isArray(input.requested_families)
      ? input.requested_families.length
      : 0;

  const wrestleInput: ResearchWrestleSessionInput = {
    session_id,
    parent_asset_id,
    floating_instance_count: input.floating_instance_count,
    completed_floating_count: input.completed_floating_count,
    twin_insight_count: input.twin_insight_count,
    twin_question_count: input.twin_question_count,
    open_question_count: input.open_question_count,
    source_family_count: familyCount,
    citation_pack_ready: competition_quality.citations.pack_ready,
    quality_overall: input.quality_overall,
    quality_floor: input.quality_floor,
    would_exceed: input.would_exceed,
    preferred_view_mode: input.preferred_view_mode,
    operator_override: input.operator_override,
  };
  const wrestle = composeResearchWrestleSession(wrestleInput);
  notes.push(...wrestle.notes);

  const session_ready =
    wrestle.wrestle_ready &&
    competition_quality.pack_ready &&
    wrestle.live_dispatch_authorized === false &&
    competition_quality.live_dispatch_authorized === false;

  if (!wrestle.wrestle_ready) {
    notes.push("session_ready=false — wrestle substrate not ready");
  } else if (!competition_quality.pack_ready) {
    notes.push(
      "session_ready=false — competition/quality/source pack not ready",
    );
  } else {
    notes.push(
      "session_ready=true — wrestle+competition quality intent only; still pure",
    );
  }

  if (
    wrestle.live_dispatch_authorized !== false ||
    competition_quality.live_dispatch_authorized !== false ||
    competition_quality.remote_fetched !== false ||
    competition_quality.backlog_mutated !== false
  ) {
    throw new Error("invariant: nested honesty flags must remain false");
  }

  notes.push("live_dispatch_authorized=false");
  notes.push("remote_fetched=false");
  notes.push("backlog_mutated=false");

  return {
    session_id,
    parent_asset_id,
    wrestle,
    competition_quality,
    session_ready,
    live_dispatch_authorized: false,
    remote_fetched: false,
    backlog_mutated: false,
    notes,
    authority: "research_wrestle_competition_quality_compose_advisory",
  };
}

export function formatResearchWrestleCompetitionQualitySummary(
  c: ResearchWrestleCompetitionQualityCompose,
): string {
  return (
    `session_ready=${c.session_ready} · wrestle_ready=${c.wrestle.wrestle_ready} · ` +
    `quality_pack_ready=${c.competition_quality.pack_ready} · ` +
    `citations=${c.competition_quality.citations.citation_count} · ` +
    `behind=${c.competition_quality.competition.behind_count} · ` +
    `live_dispatch_authorized=false · remote_fetched=false · backlog_mutated=false`
  );
}
